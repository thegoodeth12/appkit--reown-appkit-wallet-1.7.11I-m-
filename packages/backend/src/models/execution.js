import { DateTime } from 'luxon';
import Base from '@/models/base.js';
import Flow from '@/models/flow.js';
import ExecutionStep from '@/models/execution-step.js';
import Telemetry from '@/helpers/telemetry/index.js';

class Execution extends Base {
  static tableName = 'executions';

  static jsonSchema = {
    type: 'object',

    properties: {
      id: { type: 'string', format: 'uuid' },
      flowId: { type: 'string', format: 'uuid' },
      testRun: { type: 'boolean', default: false },
      internalId: { type: 'string' },
      status: { type: 'string', enum: ['success', 'failure'] },
      deletedAt: { type: 'string' },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    flow: {
      relation: Base.BelongsToOneRelation,
      modelClass: Flow,
      join: {
        from: 'executions.flow_id',
        to: 'flows.id',
      },
    },
    executionSteps: {
      relation: Base.HasManyRelation,
      modelClass: ExecutionStep,
      join: {
        from: 'executions.id',
        to: 'execution_steps.execution_id',
      },
    },
  });

  static find({ name, status, startDateTime, endDateTime }) {
    return this.query()
      .withSoftDeleted()
      .joinRelated({
        flow: true,
      })
      .withGraphFetched({
        flow: {
          steps: true,
        },
      })
      .where((builder) => {
        builder.withSoftDeleted();

        if (name) {
          builder.where('flow.name', 'ilike', `%${name}%`);
        }

        if (status === 'success') {
          builder.where('executions.status', 'success');
        } else if (status === 'failure') {
          builder.where('executions.status', 'failure');
        }

        if (startDateTime) {
          const startDate = DateTime.fromMillis(Number(startDateTime));

          if (startDate.isValid) {
            builder.where('executions.created_at', '>=', startDate.toISO());
          }
        }

        if (endDateTime) {
          const endDate = DateTime.fromMillis(Number(endDateTime));

          if (endDate.isValid) {
            builder.where('executions.created_at', '<=', endDate.toISO());
          }
        }
      })
      .orderBy('created_at', 'desc');
  }

  async $afterInsert(queryContext) {
    await super.$afterInsert(queryContext);
    Telemetry.executionCreated(this);
  }
}

export default Execution;
