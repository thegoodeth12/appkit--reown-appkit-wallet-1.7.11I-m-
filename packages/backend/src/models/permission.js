import Base from '@/models/base.js';
import permissionCatalog from '@/helpers/permission-catalog.ee.js';

class Permission extends Base {
  static tableName = 'permissions';

  static jsonSchema = {
    type: 'object',
    required: ['roleId', 'action', 'subject'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      roleId: { type: 'string', format: 'uuid' },
      action: { type: 'string', minLength: 1 },
      subject: { type: 'string', minLength: 1 },
      conditions: { type: 'array', items: { type: 'string' } },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static filter(permissions) {
    const sanitizedPermissions = permissions.filter((permission) => {
      const { action, subject, conditions } = permission;

      const relevantAction = this.findAction(action);
      const validSubject = this.isSubjectValid(subject, relevantAction);
      const validConditions = this.areConditionsValid(conditions);

      return relevantAction && validSubject && validConditions;
    });

    return sanitizedPermissions;
  }

  static findAction(action) {
    return permissionCatalog.actions.find(
      (actionCatalogItem) => actionCatalogItem.key === action
    );
  }

  static isSubjectValid(subject, action) {
    return action && action.subjects.includes(subject);
  }

  static areConditionsValid(conditions) {
    return conditions.every((condition) => this.isConditionValid(condition));
  }

  static isConditionValid(condition) {
    return !!permissionCatalog.conditions.find(
      (conditionCatalogItem) => conditionCatalogItem.key === condition
    );
  }
}

export default Permission;
