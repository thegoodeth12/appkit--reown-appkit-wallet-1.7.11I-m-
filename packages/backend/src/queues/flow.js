import { generateQueue } from '@/queues/queue.js';

const flowQueue = generateQueue('flow');
export default flowQueue;
