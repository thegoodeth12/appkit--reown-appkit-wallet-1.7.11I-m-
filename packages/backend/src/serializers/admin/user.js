import userSerializer from '@/serializers/user.js';

const adminUserSerializer = (user) => {
  const userData = userSerializer(user);

  userData.acceptInvitationUrl = user.acceptInvitationUrl;

  return userData;
};

export default adminUserSerializer;
