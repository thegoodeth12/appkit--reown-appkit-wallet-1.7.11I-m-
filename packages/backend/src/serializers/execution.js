import flowSerializer from '@/serializers/flow.js';

const executionSerializer = (execution) => {
  let executionData = {
    id: execution.id,
    testRun: execution.testRun,
    status: execution.status,
    createdAt: execution.createdAt.getTime(),
    updatedAt: execution.updatedAt.getTime(),
  };

  if (execution.status) {
    executionData.status = execution.status;
  }

  if (execution.flow) {
    executionData.flow = flowSerializer(execution.flow);
  }

  return executionData;
};

export default executionSerializer;
