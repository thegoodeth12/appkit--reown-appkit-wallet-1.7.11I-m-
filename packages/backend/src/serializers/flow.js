import stepSerializer from '@/serializers/step.js';
import folderSerilializer from '@/serializers/folder.js';

const flowSerializer = (flow) => {
  let flowData = {
    id: flow.id,
    name: flow.name,
    active: flow.active,
    status: flow.status,
    createdAt: flow.createdAt.getTime(),
    updatedAt: flow.updatedAt.getTime(),
  };

  if ('isOwner' in flow) {
    flowData.isOwner = flow.isOwner;
  }

  if (flow.steps?.length > 0) {
    flowData.steps = flow.steps.map((step) => stepSerializer(step));
  }

  if (flow.folder) {
    flowData.folder = folderSerilializer(flow.folder);
  }

  return flowData;
};

export default flowSerializer;
