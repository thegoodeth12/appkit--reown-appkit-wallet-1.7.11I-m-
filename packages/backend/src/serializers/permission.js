const permissionSerializer = (permission) => {
  return {
    id: permission.id,
    roleId: permission.roleId,
    action: permission.action,
    subject: permission.subject,
    conditions: permission.conditions,
    createdAt: permission.createdAt.getTime(),
    updatedAt: permission.updatedAt.getTime(),
  };
};

export default permissionSerializer;
