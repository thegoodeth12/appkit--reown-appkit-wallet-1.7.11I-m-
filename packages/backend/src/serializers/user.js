import roleSerializer from '@/serializers/role.js';
import permissionSerializer from '@/serializers/permission.js';
import appConfig from '@/config/app.js';

const userSerializer = (user) => {
  let userData = {
    id: user.id,
    email: user.email,
    createdAt: user.createdAt.getTime(),
    updatedAt: user.updatedAt.getTime(),
    status: user.status,
    fullName: user.fullName,
  };

  if (user.role) {
    userData.role = roleSerializer(user.role);
  }

  if (user.permissions?.length > 0) {
    userData.permissions = user.permissions.map((permission) =>
      permissionSerializer(permission)
    );
  }

  if (appConfig.isCloud && user.trialExpiryDate) {
    userData.trialExpiryDate = user.trialExpiryDate;
  }

  return userData;
};

export default userSerializer;
