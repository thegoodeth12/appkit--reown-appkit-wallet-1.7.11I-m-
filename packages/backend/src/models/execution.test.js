import { vi, describe, it, expect, beforeEach } from 'vitest';
import Execution from '@/models/execution.js';
import ExecutionStep from '@/models/execution-step.js';
import Flow from '@/models/flow.js';
import Base from '@/models/base.js';
import Telemetry from '@/helpers/telemetry/index.js';
import { createExecution } from '@/factories/execution.js';
import { createUser } from '@/factories/user.js';
import { createFlow } from '@/factories/flow.js';
import { createStep } from '@/factories/step.js';
import { createPermission } from '@/factories/permission.js';

describe('Execution model', () => {
  it('tableName should return correct name', () => {
    expect(Execution.tableName).toBe('executions');
  });

  it('jsonSchema should have correct validations', () => {
    expect(Execution.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = Execution.relationMappings();

    const expectedRelations = {
      executionSteps: {
        join: {
          from: 'executions.id',
          to: 'execution_steps.execution_id',
        },
        modelClass: ExecutionStep,
        relation: Base.HasManyRelation,
      },
      flow: {
        join: {
          from: 'executions.flow_id',
          to: 'flows.id',
        },
        modelClass: Flow,
        relation: Base.BelongsToOneRelation,
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  describe('find', () => {
    let currentUser,
      currentUserRole,
      anotherUser,
      flow,
      executionOne,
      executionTwo,
      executionThree;

    beforeEach(async () => {
      currentUser = await createUser();
      currentUserRole = await currentUser.$relatedQuery('role');

      anotherUser = await createUser();

      flow = await createFlow({
        userId: currentUser.id,
        name: 'Test Flow',
      });

      const anotherUserFlow = await createFlow({
        userId: anotherUser.id,
        name: 'Another User Flow',
      });

      executionOne = await createExecution({
        flowId: flow.id,
        testRun: false,
        status: 'success',
      });

      // sleep for 10 milliseconds to make sure the created_at values are different
      await new Promise((resolve) => setTimeout(resolve, 10));

      executionTwo = await createExecution({
        flowId: flow.id,
        testRun: true,
        status: 'failure',
      });

      // sleep for 10 milliseconds to make sure the created_at values are different
      await new Promise((resolve) => setTimeout(resolve, 10));

      executionThree = await createExecution({
        flowId: anotherUserFlow.id,
        testRun: false,
        status: 'success',
      });

      await createPermission({
        action: 'read',
        subject: 'Execution',
        roleId: currentUserRole.id,
        conditions: [],
      });

      currentUser = await currentUser.$query().withGraphFetched({
        role: true,
        permissions: true,
      });
    });

    it('should return executions filtered by name', async () => {
      const executions = await Execution.find({ name: 'Test Flow' });

      expect(executions).toHaveLength(2);

      expect(executions[0].id).toBe(executionTwo.id);
      expect(executions[1].id).toBe(executionOne.id);
    });

    it('should return executions filtered by success status', async () => {
      const executions = await Execution.find({ status: 'success' });

      expect(executions).toHaveLength(2);

      expect(executions[0].id).toBe(executionThree.id);
      expect(executions[1].id).toBe(executionOne.id);
    });

    it('should return executions filtered by failure status', async () => {
      const executions = await Execution.find({ status: 'failure' });

      expect(executions).toHaveLength(1);
      expect(executions[0].id).toBe(executionTwo.id);
    });

    it('should return executions filtered by startDateTime and endDateTime', async () => {
      const executions = await Execution.find({
        startDateTime: executionOne.createdAt,
        endDateTime: executionTwo.createdAt,
      });

      expect(executions).toHaveLength(2);
      expect(executions[0].id).toBe(executionTwo.id);
      expect(executions[1].id).toBe(executionOne.id);
    });

    it('should return all executions when no filter is applied', async () => {
      const executions = await Execution.find({});

      expect(executions.length).toBeGreaterThanOrEqual(3);

      expect(executions.some((e) => e.id === executionOne.id)).toBe(true);
      expect(executions.some((e) => e.id === executionTwo.id)).toBe(true);
      expect(executions.some((e) => e.id === executionThree.id)).toBe(true);
    });

    it('should include flow and steps in the returned executions', async () => {
      const step = await createStep({
        flowId: flow.id,
        type: 'trigger',
      });

      const executions = await Execution.find({ name: 'Test Flow' });

      expect(executions[0].flow.id).toBe(flow.id);
      expect(executions[0].flow.steps[0].id).toBe(step.id);
    });
  });

  it('$afterInsert should call Telemetry.executionCreated', async () => {
    const telemetryExecutionCreatedSpy = vi
      .spyOn(Telemetry, 'executionCreated')
      .mockImplementation(() => {});

    const execution = await createExecution();

    expect(telemetryExecutionCreatedSpy).toHaveBeenCalledWith(execution);
  });
});
