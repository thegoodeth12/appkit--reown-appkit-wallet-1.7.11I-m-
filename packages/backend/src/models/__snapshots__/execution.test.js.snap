// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Execution model > jsonSchema should have correct validations 1`] = `
{
  "properties": {
    "createdAt": {
      "type": "string",
    },
    "deletedAt": {
      "type": "string",
    },
    "flowId": {
      "format": "uuid",
      "type": "string",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "internalId": {
      "type": "string",
    },
    "status": {
      "enum": [
        "success",
        "failure",
      ],
      "type": "string",
    },
    "testRun": {
      "default": false,
      "type": "boolean",
    },
    "updatedAt": {
      "type": "string",
    },
  },
  "type": "object",
}
`;
